# SVM ML LAB

### Zaidan Mufaddhal | AIU20092029

##  1. Data Preprocessing

In [34]:
# importing libraries  
import numpy as nm  
import matplotlib.pyplot as mtp  
import pandas as pd 

In [35]:
#importing datasets  
df = pd.read_csv('Downloads/MSFT_forSVM.csv')
df

Date        Open        High         Low       Close   Adj Close  \
0     2017/05/19   67.500000   68.099998   67.430000   67.690002   63.549782   
1     2017/05/22   67.889999   68.500000   67.500000   68.449997   64.263275   
2     2017/05/23   68.720001   68.750000   68.379997   68.680000   64.479210   
3     2017/05/24   68.870003   68.879997   68.449997   68.769997   64.563705   
4     2017/05/25   68.970001   69.879997   68.910004   69.620003   65.361717   
...          ...         ...         ...         ...         ...         ...   
1254  2022/05/12  257.690002  259.880005  250.020004  255.350006  255.350006   
1255  2022/05/13  257.350006  263.040009  255.350006  261.119995  261.119995   
1256  2022/05/16  259.959991  265.820007  255.779999  261.500000  261.500000   
1257  2022/05/17  266.109985  268.329987  262.459991  266.820007  266.820007   
1258  2022/05/18  263.000000  263.600006  256.410004  257.019989  257.019989   

        Volume  
0     26961100  
1     16237600  
2     15425800  
3     14593900  
4     21854100  
...        ...  
1254  51033800  
1255  34893400  
1256  32550900  
1257  28806100  
1258  11705360  

[1259 rows x 7 columns]

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       1259 non-null   object 
 1   Open       1259 non-null   float64
 2   High       1259 non-null   float64
 3   Low        1259 non-null   float64
 4   Close      1259 non-null   float64
 5   Adj Close  1259 non-null   float64
 6   Volume     1259 non-null   int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 69.0+ KB


In [37]:
df.describe()

Open         High          Low        Close    Adj Close  \
count  1259.000000  1259.000000  1259.000000  1259.000000  1259.000000   
mean    171.869913   173.563701   170.030064   171.880064   168.859214   
std      79.573595    80.379943    78.653997    79.546935    80.789037   
min      67.500000    68.099998    67.430000    67.690002    63.549782   
25%     104.115002   105.449997   102.625000   104.230000   100.195797   
50%     144.970001   147.500000   143.759995   145.960007   142.449112   
75%     234.485001   236.849998   231.620002   234.189995   232.297837   
max     344.619995   349.670013   342.200012   343.109985   342.402008   

             Volume  
count  1.259000e+03  
mean   2.947084e+07  
std    1.307347e+07  
min    7.425600e+06  
25%    2.124435e+07  
50%    2.613270e+07  
75%    3.365610e+07  
max    1.112421e+08

#### Get Independent and dependent data

In [38]:
#Create empty lists to store the independent and dependent data
days = list()
adj_close_prices = list()

In [39]:
#Get the dates and lists to store independent and dependent data
df_days = df.loc[:, 'Date']
df_adj_price = df.loc[:, 'Close']

In [40]:
#Create the independent data set
for day in df_days:
    days.append( [int(day.split('/')[1])])
    
#Create the dependent data set
for adj_close_price in df_adj_price:
    adj_close_prices.append(float(adj_close_price))

In [41]:
x = days
y = adj_close_prices

In [42]:
# Splitting the dataset into training and test set.  
from sklearn.model_selection import train_test_split  
x_train, x_test, y_train, y_test= train_test_split(x, y, test_size= 0.3, random_state=0)  
#feature Scaling  
from sklearn.preprocessing import StandardScaler    
st_x= StandardScaler()    
x_train= st_x.fit_transform(x_train)    
x_test= st_x.transform(x_test) 

In [43]:
x_train

array([[ 0.39956868],
       [ 1.55657773],
       [ 0.68882094],
       [ 1.26732546],
       [ 0.9780732 ],
       [-0.75744037],
       [ 1.26732546],
       [-1.04669264],
       [-0.46818811],
       [ 0.39956868],
       [-1.62519716],
       [ 0.68882094],
       [ 0.68882094],
       [-0.17893585],
       [ 0.68882094],
       [ 0.11031641],
       [-1.62519716],
       [ 0.11031641],
       [-1.62519716],
       [ 0.39956868],
       [-1.3359449 ],
       [-1.62519716],
       [ 1.26732546],
       [ 0.9780732 ],
       [-1.04669264],
       [-1.3359449 ],
       [ 0.11031641],
       [-0.75744037],
       [-0.17893585],
       [-1.04669264],
       [-0.17893585],
       [ 0.39956868],
       [-1.04669264],
       [ 1.55657773],
       [ 0.68882094],
       [ 0.9780732 ],
       [-0.17893585],
       [-0.17893585],
       [-0.17893585],
       [-0.75744037],
       [ 0.11031641],
       [-0.17893585],
       [ 1.55657773],
       [-1.04669264],
       [ 1.26732546],
       [-0

In [44]:
x_test

array([[-0.46818811],
       [-0.46818811],
       [ 0.39956868],
       [-0.75744037],
       [-1.3359449 ],
       [-0.17893585],
       [ 0.39956868],
       [-0.17893585],
       [ 0.9780732 ],
       [ 0.39956868],
       [ 0.9780732 ],
       [-1.04669264],
       [-1.62519716],
       [-1.04669264],
       [-0.46818811],
       [-1.04669264],
       [-1.62519716],
       [ 1.55657773],
       [ 0.68882094],
       [-0.46818811],
       [-1.62519716],
       [ 1.26732546],
       [-1.04669264],
       [-0.46818811],
       [ 1.55657773],
       [-1.04669264],
       [ 1.55657773],
       [ 0.11031641],
       [-1.04669264],
       [ 0.39956868],
       [-1.62519716],
       [ 0.39956868],
       [ 0.9780732 ],
       [ 0.11031641],
       [ 1.55657773],
       [ 1.26732546],
       [ 0.39956868],
       [ 1.26732546],
       [-0.75744037],
       [ 0.39956868],
       [-0.17893585],
       [-0.75744037],
       [-0.17893585],
       [-1.3359449 ],
       [ 1.26732546],
       [-0

In [45]:
y_test

[69.959999,
 125.510002,
 72.150002,
 260.579987,
 90.809998,
 69.900002,
 107.059998,
 119.839996,
 103.849998,
 135.740005,
 138.429993,
 161.570007,
 224.970001,
 120.220001,
 174.570007,
 232.419998,
 224.339996,
 341.25,
 74.940002,
 274.730011,
 232.330002,
 84.050003,
 135.419998,
 128.149994,
 342.450012,
 89.790001,
 222.75,
 289.049988,
 112.830002,
 208.25,
 88.220001,
 209.190002,
 294.850006,
 74.190002,
 219.279999,
 147.070007,
 209.699997,
 151.229996,
 123.370003,
 109.669998,
 101.419998,
 175.059998,
 253.809998,
 295.0,
 216.550003,
 68.449997,
 211.75,
 208.699997,
 84.160004,
 183.509995,
 227.389999,
 94.07,
 333.100006,
 72.790001,
 151.75,
 113.370003,
 163.490005,
 310.109985,
 73.349998,
 85.400002,
 106.279999,
 208.350006,
 301.149994,
 111.589996,
 94.129997,
 219.419998,
 237.580002,
 139.100006,
 282.51001,
 299.869995,
 85.949997,
 89.129997,
 219.660004,
 70.32,
 85.230003,
 140.399994,
 155.710007,
 105.010002,
 106.120003,
 252.460007,
 98.660004,
 1

In [46]:
y_train

[228.910004,
 223.940002,
 137.520004,
 105.919998,
 142.830002,
 260.73999,
 107.139999,
 93.860001,
 126.900002,
 73.160004,
 106.199997,
 137.320007,
 205.369995,
 100.860001,
 294.799988,
 139.289993,
 88.349998,
 204.699997,
 216.440002,
 135.279999,
 244.490005,
 212.649994,
 335.950012,
 136.369995,
 285.589996,
 92.0,
 73.300003,
 153.830002,
 185.360001,
 150.619995,
 247.399994,
 105.980003,
 89.470001,
 323.799988,
 300.209991,
 108.300003,
 257.380005,
 100.790001,
 70.0,
 120.769997,
 68.169998,
 123.160004,
 216.009995,
 299.160004,
 202.330002,
 97.5,
 162.089996,
 216.470001,
 137.380005,
 153.240005,
 210.520004,
 237.039993,
 100.879997,
 106.959999,
 341.269989,
 77.120003,
 221.020004,
 139.440002,
 157.699997,
 209.830002,
 106.870003,
 158.669998,
 137.410004,
 316.380005,
 149.479996,
 72.400002,
 178.589996,
 139.139999,
 336.070007,
 247.860001,
 336.630005,
 146.919998,
 73.160004,
 231.649994,
 125.5,
 137.350006,
 165.270004,
 214.199997,
 97.540001,
 202.67

## 2. Fitting the SVM classifier to the training set

In [50]:
from sklearn.svm import SVR # "Support vector classifier"  

In [53]:
classifier = SVR(kernel='linear', C=1000.0) 

In [54]:
classifier.fit(x_train, y_train) 

SVR(C=1000.0, kernel='linear')

## 3. Predicting the test set result

In [55]:
y_pred=classifier.predict(x_test)

In [56]:
y_test

[69.959999,
 125.510002,
 72.150002,
 260.579987,
 90.809998,
 69.900002,
 107.059998,
 119.839996,
 103.849998,
 135.740005,
 138.429993,
 161.570007,
 224.970001,
 120.220001,
 174.570007,
 232.419998,
 224.339996,
 341.25,
 74.940002,
 274.730011,
 232.330002,
 84.050003,
 135.419998,
 128.149994,
 342.450012,
 89.790001,
 222.75,
 289.049988,
 112.830002,
 208.25,
 88.220001,
 209.190002,
 294.850006,
 74.190002,
 219.279999,
 147.070007,
 209.699997,
 151.229996,
 123.370003,
 109.669998,
 101.419998,
 175.059998,
 253.809998,
 295.0,
 216.550003,
 68.449997,
 211.75,
 208.699997,
 84.160004,
 183.509995,
 227.389999,
 94.07,
 333.100006,
 72.790001,
 151.75,
 113.370003,
 163.490005,
 310.109985,
 73.349998,
 85.400002,
 106.279999,
 208.350006,
 301.149994,
 111.589996,
 94.129997,
 219.419998,
 237.580002,
 139.100006,
 282.51001,
 299.869995,
 85.949997,
 89.129997,
 219.660004,
 70.32,
 85.230003,
 140.399994,
 155.710007,
 105.010002,
 106.120003,
 252.460007,
 98.660004,
 1

In [57]:
y_pred

array([150.64944766, 150.64944766, 142.14201754, 153.48525771,
       159.15687779, 147.81363762, 142.14201754, 147.81363762,
       136.47039745, 142.14201754, 136.47039745, 156.32106775,
       161.99268783, 156.32106775, 150.64944766, 156.32106775,
       161.99268783, 130.79877737, 139.3062075 , 150.64944766,
       161.99268783, 133.63458741, 156.32106775, 150.64944766,
       130.79877737, 156.32106775, 130.79877737, 144.97782758,
       156.32106775, 142.14201754, 161.99268783, 142.14201754,
       136.47039745, 144.97782758, 130.79877737, 133.63458741,
       142.14201754, 133.63458741, 153.48525771, 142.14201754,
       147.81363762, 153.48525771, 147.81363762, 159.15687779,
       133.63458741, 150.64944766, 144.97782758, 142.14201754,
       130.79877737, 150.64944766, 156.32106775, 150.64944766,
       130.79877737, 142.14201754, 130.79877737, 139.3062075 ,
       153.48525771, 136.47039745, 144.97782758, 130.79877737,
       142.14201754, 144.97782758, 139.3062075 , 159.15

## 4. Creating the confusion matrix

In [58]:
from sklearn.metrics import confusion_matrix  
cm= confusion_matrix(y_test, y_pred)

ValueError: continuous is not supported

In [60]:
from sklearn.metrics import accuracy_score
r2_score(y_test, y_pred)

NameError: name 'r2_score' is not defined

In [40]:
cm

array([[141,  16],
       [ 34,  40]], dtype=int64)

So, based on the above result, there is 141+40=181 == 78.35% correct predictions and 34+16=50 == 21.64% incorrect predictions.

## 5. Visualizing the training set result

In [61]:
from matplotlib.colors import ListedColormap  

In [62]:
x_set, y_set = x_train, y_train  
x1, x2 = nm.meshgrid(nm.arange(start = x_set[:, 0].min() - 1, stop = x_set[:, 0].max() + 1, step  =0.01),  
                     nm.arange(start = x_set[:, 1].min() - 1, stop = x_set[:, 1].max() + 1, step = 0.01)) 
mtp.contourf(x1, x2, classifier.predict(nm.array([x1.ravel(), x2.ravel()] + [nm.repeat(0, x1.ravel().size) for _ in range(6)]).T).reshape(x1.shape),  
    alpha = 0.75, cmap = ListedColormap(('red', 'green')))  
mtp.xlim(x1.min(), x1.max())  
mtp.ylim(x2.min(), x2.max())  
for i, j in enumerate(nm.unique(y_set)):  
    mtp.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],  
        c = ListedColormap(('red', 'green'))(i), label = j)  
mtp.title('SVM classifier (Training set)')  
mtp.xlabel('Age')  
mtp.ylabel('Medical Predictor Variables')  
mtp.legend()  
mtp.show()

IndexError: index 1 is out of bounds for axis 1 with size 1

## 6. Visualizing the test set result

In [63]:
x_set, y_set = x_test, y_test  
x1, x2 = nm.meshgrid(nm.arange(start = x_set[:, 0].min() - 1, stop = x_set[:, 0].max() + 1, step  =0.01),  
                     nm.arange(start = x_set[:, 1].min() - 1, stop = x_set[:, 1].max() + 1, step = 0.01)) 
mtp.contourf(x1, x2, classifier.predict(nm.array([x1.ravel(), x2.ravel()] + [nm.repeat(0, x1.ravel().size) for _ in range(6)]).T).reshape(x1.shape),  
    alpha = 0.75, cmap = ListedColormap(('red', 'green')))  
mtp.xlim(x1.min(), x1.max())  
mtp.ylim(x2.min(), x2.max())  
for i, j in enumerate(nm.unique(y_set)):  
    mtp.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],  
        c = ListedColormap(('red', 'green'))(i), label = j)  
mtp.title('SVM classifier (Test set)')  
mtp.xlabel('Age')  
mtp.ylabel('Medical Predictor Variables')  
mtp.legend()  
mtp.show()

IndexError: index 1 is out of bounds for axis 1 with size 1